src/main/java/org/elasticsearch/script/ScriptModes.java

```diff
@@ -0,0 +1,202 @@
+
+        //file scripts are enabled by default, for any language
+        addGlobalScriptTypeModes(scriptEngines.keySet(), ScriptType.FILE, ScriptMode.ENABLE, scriptModesMap);
+        //indexed scripts are enabled by default only for sandboxed languages
+        addGlobalScriptTypeModes(scriptEngines.keySet(), ScriptType.INDEXED, ScriptMode.SANDBOX, scriptModesMap);
+        //dynamic scripts are enabled by default only for sandboxed languages
+        addGlobalScriptTypeModes(scriptEngines.keySet(), ScriptType.INLINE, ScriptMode.SANDBOX, scriptModesMap);
+
+        boolean newScriptSettingsUsed = false;
+
+        //read custom source based settings for all operations (e.g. script.indexed: enable)
+        for (ScriptType scriptType : ScriptType.values()) {
+            String scriptTypeSetting = settings.get(SCRIPT_SETTINGS_PREFIX + scriptType);
+            if (Strings.hasLength(scriptTypeSetting)) {
+                newScriptSettingsUsed = true;
+                ScriptMode scriptTypeMode = ScriptMode.parse(scriptTypeSetting);
+                addGlobalScriptTypeModes(scriptEngines.keySet(), scriptType, scriptTypeMode, scriptModesMap);
+            }
+        }
+
+        //read custom op based settings for all sources (e.g. script.aggs: disable)
+        //op based settings take precedence over source based settings, hence they get expanded later
+        for (ScriptedOp scriptedOp : ScriptedOp.values()) {
+            ScriptMode scriptMode = getScriptedOpMode(settings, SCRIPT_SETTINGS_PREFIX, scriptedOp);
+            if (scriptMode != null) {
+                newScriptSettingsUsed = true;
+                addGlobalScriptedOpModes(scriptEngines.keySet(), scriptedOp, scriptMode, scriptModesMap);
+            }
+        }
+
+        Map<String, Settings> langGroupedSettings = settings.getGroups(ENGINE_SETTINGS_PREFIX, true);
+        for (Map.Entry<String, Settings> langSettings : langGroupedSettings.entrySet()) {
+            //read engine specific settings that refer to a non existing script lang will be ignored
+            ScriptEngineService scriptEngineService = scriptEngines.get(langSettings.getKey());
+            if (scriptEngineService != null) {
+                for (ScriptType scriptType : ScriptType.values()) {
+                    for (ScriptedOp scriptedOp : ScriptedOp.values()) {
+                        ScriptMode scriptMode = getScriptedOpMode(langSettings.getValue(), scriptType + ".", scriptedOp);
+                        if (scriptMode != null) {
+                            newScriptSettingsUsed = true;
+                            addScriptMode(scriptEngineService, scriptType, scriptedOp, scriptMode, scriptModesMap);
+                        }
+                    }
+                }
+            }
+        }
+
+        //read deprecated disable_dynamic setting, apply only if none of the new settings is used
+        String disableDynamicSetting = settings.get(ScriptService.DISABLE_DYNAMIC_SCRIPTING_SETTING);
```

Comment:
I broke it up in different methods that process their own part fo the settings in order. That said I am not sure about testing. I think how we currently test this class is pretty good and testing these methods separately would create some duplication, not sure it's really worth it. At the end of the day the coverage here is already high and the amount of test code is quite big too.